### PR TITLE
Add iidfile flag for build image

### PR DESCRIPTION
**Pull request by Junnplus**
related https://github.com/containerd/nerdctl/issues/215

### CODE REVIEW

File: cmd/nerdctl/build.go [hunk 1]
```diff
@@ -146,6 +149,17 @@ func buildAction(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
+	iidFile, _ := cmd.Flags().GetString("iidfile")
+	if iidFile != "" {
+		id, err := getDigestFromMetaFile(iidFile)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:154]: ID can be fetched from loadImage() with a modification like this https://github.com/containerd/nerdctl/pull/582/files#diff-714eb9b930b7e765168c40536f11caba9f7c5e8c9d0f097c2c0da77842a53ae6R183

File: cmd/nerdctl/build.go [hunk 2]
```diff
@@ -146,6 +149,17 @@ func buildAction(cmd *cobra.Command, args []string) error {
...
+		if err != nil {
+			return err
+		}
+		if err := writeIIDFile(iidFile, id); err != nil {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:158]: os.WriteFile

File: cmd/nerdctl/build.go [hunk 3]
```diff
@@ -17,9 +17,11 @@
 package main
 
 import (
+	"encoding/json"
 	"errors"
 	"fmt"
 	"io"
+	"io/ioutil"
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:24]: ioutil is deprecated

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: cmd/nerdctl/build.go (hunk 1 above)
Comment (Junnplus) [reply cmd/nerdctl/build.go:154]: `--output=type=image --iidfile=id`

`loadImage` should not be called when `output=type=image`, will raise `unrecognized image format` error.

File: cmd/nerdctl/build.go [hunk 4]
```diff
@@ -318,5 +331,36 @@ func generateBuildctlArgs(cmd *cobra.Command, platform, args []string) (string,
 		buildctlArgs = append(buildctlArgs, "--export-cache="+s)
 	}
 
+	iidFile, err := cmd.Flags().GetString("iidfile")
+	if err != nil {
+		return "", nil, false, cleanup, err
+	}
+	if iidFile != "" {
+		buildctlArgs = append(buildctlArgs, "--metadata-file="+iidFile)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:339]: Confusing to write metadata file to iidfile.
Better to create a metadata file in `os.MkdirTemp`

File: cmd/nerdctl/build.go [hunk 5]
```diff
@@ -318,5 +331,36 @@ func generateBuildctlArgs(cmd *cobra.Command, platform, args []string) (string,
...
+	}
+
 	return buildctlBinary, buildctlArgs, needsLoading, cleanup, nil
 }
+
+func getDigestFromMetaFile(path string) (string, error) {
+	f, err := os.Open(path)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:346]: os.ReadFile

**Comment by ktock** [comment]
Got a different result than docker after some manual testing.

docker:

```
$ docker build -t test --iidfile=/tmp/iid /tmp/ctx
$ cat /tmp/iid
sha256:c74973c38ea79154adb9c8caf29e4be8f9b09690f08cd36bee480b3568ba2da5
$ docker image inspect test | jq -r '.[0].Id'
sha256:c74973c38ea79154adb9c8caf29e4be8f9b09690f08cd36bee480b3568ba2da5
```

nerdctl

```
# nerdctl build -t test --iidfile=/tmp/iid /tmp/ctx
# cat /tmp/iid
sha256:84199e0a1de5a053170744bd4b7d6333480ada680136a9fed0e811e409fc1946
# nerdctl image inspect test | jq -r '.[0].Id'
sha256:c7a9088f63ec485c116a91e5f6834c2dba5e55da5a390b87a8864bcde9778432
```

**Comment by AkihiroSuda** [comment]
> iidfile should be that same as nerdctl image inspect at least.

No.
https://github.com/containerd/nerdctl/pull/442

**Comment by ktock** [comment]
LGTM on green

**Comment by AkihiroSuda** [comment]
Added a follow-up commit for docs

**Review by ktock — APPROVED** [review]

**Pull request merged**